File: src/robot_mqtt/src/rtmp_streamer.cpp
#include <ros/ros.h>
#include <cv_bridge/cv_bridge.h>
#include <sensor_msgs/Image.h>
#include <opencv2/core.hpp>

class RTMPStreamer {
public:
    RTMPStreamer() {
        // 初始化 FFmpeg 管道（示例参数，需调整）
        ffmpeg_cmd = "ffmpeg -f rawvideo -pix_fmt bgr24 -s 1280x760 -r 30 -i - -c:v libx264 -f flv rtmp://tx.direct.huya.com/huyalive/1199574560753-1199574560753-7484892029064985869-2399149244962-10057-A-1742712467-1?seq=1745919923386&type=simple";
        ffmpeg_proc = popen(ffmpeg_cmd.c_str(), "w");
        sub = nh.subscribe("/camera/color/image_raw", 10, &RTMPStreamer::imageCallback, this);
    }

    void imageCallback(const sensor_msgs::ImageConstPtr& msg) {
        try {
            cv_bridge::CvImagePtr cv_ptr = cv_bridge::toCvCopy(msg, "bgr8");
            fwrite(cv_ptr->image.data, 1, cv_ptr->image.total() * cv_ptr->image.elemSize(), ffmpeg_proc);
        } catch (cv_bridge::Exception& e) {
            ROS_ERROR("cv_bridge exception: %s", e.what());
        }
    }

private:
    ros::NodeHandle nh;
    ros::Subscriber sub;
    FILE* ffmpeg_proc;
    std::string ffmpeg_cmd;
};

int main(int argc, char** argv) {
    ros::init(argc, argv, "rtmp_streamer");
    RTMPStreamer streamer;
    ros::spin();
    return 0;
}

File: src/robot_mqtt/scripts/rtmp_streamer.py
#!/usr/bin/env python3
import rospy
import subprocess
from sensor_msgs.msg import Image
import cv2
from cv_bridge import CvBridge

class RTMPStreamer:
    def __init__(self):
        self.bridge = CvBridge()
        self.ffmpeg_cmd = [
            'ffmpeg',
            '-f', 'rawvideo',      # 输入格式为原始视频
            '-pix_fmt', 'bgr24',   # OpenCV默认格式为BGR
            '-s', '640x480',       # 分辨率（需与图像一致）
            '-r', '30',            # 帧率
            '-i', '-',             # 从标准输入读取数据
            '-c:v', 'libx264',     # 编码器
            '-preset', 'fast',     # 编码预设
            '-f', 'flv',          # 输出格式为FLV
            'rtmp://tx.direct.huya.com/huyalive/1199574560753-1199574560753-7484892029064985869-2399149244962-10057-A-1742712467-1?seq=1745919923386&type=simple'  # RTMP服务器地址
        ]
        self.ffmpeg_proc = subprocess.Popen(self.ffmpeg_cmd, stdin=subprocess.PIPE)

    def image_callback(self, msg):
        try:
            cv_image = self.bridge.imgmsg_to_cv2(msg, "bgr8")
            self.ffmpeg_proc.stdin.write(cv_image.tobytes())
        except Exception as e:
            rospy.logerr(f"Error processing image: {e}")

if __name__ == '__main__':
    rospy.init_node('rtmp_streamer')
    streamer = RTMPStreamer()
    rospy.Subscriber('/camera/color/image_raw', Image, streamer.image_callback)
    rospy.spin()